examples: added 6LoWPAN UDP example application

haukepetersen opened the pull request:
This PR supersedes #2720.
~~Rebased on #2891.~~

Markes as WIP because it is not quite working yet (at least I am not able to transfer UDP data from one node to another). Also the README.md is still missing (as I am not quite sure what needs to be done before you can send data).

So any feedback and testing is welcome!

bapclenet commented on examples/ng_6lo_udp/main.c:
```diff
@@ -0,0 +1,189 @@
+    { "udp", "send data over UDP and listen on UDP ports", udp_cmd },
+    { NULL, NULL, NULL }
+};
+
```
Maybe all those above function could have been written in a file named ``udp.c`` ?

bapclenet commented:
We can't use short address apparently ?

authmillenon commented:
I'd prefer to make the 6LoWPAN part dependent on the device.

authmillenon commented on examples/ng_6lo_udp/Makefile:
```diff
@@ -0,0 +1,29 @@
+
+# Specify the modules that are used
+USEMODULE += ng_netif
+USEMODULE += ng_sixlowpan
```
Use `ng_sixlowpan_frag`.

authmillenon replied:
And as stated below: I would make this dependent on the device.

haukepetersen replied:
no, why should it be dependent on the device? What if I want to use the radio without 6LoWPAN (which is a valid use-case...)?

authmillenon commented:
@bapclenet did you call 

```
ifconfig 6lo
```

in the shell. I would prefer that this is done automatically for IEEE 802.15.4 devices, but currently this is not included (will do a follow-up for #2891 as soon as it got merged).

bapclenet commented:
@authmillenon, yeah it `` set 6LoWPAN support on interface X`` 
Should I be able able to send via short address then? Because I still can't

authmillenon commented:
normally, yes.

OlegHahm commented:
Before we merge this, we should discuss which applications we want to have in this repository in the `examples` folder and which in the RIOT-OS/applications repository. I'm still in strong favor of keeping the number of example applications here at a minimum.

authmillenon commented:
@OlegHahm I agree with you in part. But central features as the network stack should have some example in the main repository IMHO. Mind that I'm not proposing to add an example for every little network module. I say we add this example (maybe rename it to `iot` or something like that) and add features like microcoap and RPL later on.

OlegHahm commented:
I agree that we should have an example that demonstrate the most important features of our new network stack, I'm just saying that we should not just add this example without thinking about the whole set of example applications we want to have. I guess it would make sense to extend this application with some features (like RPL or even CoAP) from the new stack.

bapclenet commented:
@haukepetersen, Please rebase

@OlegHahm, you're right about /examples. People would like to have examples with UDP and/or RPL and/or CoAP so I think, we could merge this one because this will be a good first example to show the new working network stack and further improve it with new layers. But the example should be modular. 
The previous rpl_udp_example wasn't modular and it was impossible (with simple configuration) to use UDP without RPL. Let's think about it while building this new example. This is my opinion.

OlegHahm commented:
> But the example should be modular.
The previous rpl_udp_example wasn't modular and it was impossible (with simple configuration) to use UDP without RPL. Let's think about it while building this new example. This is my opinion.

:+1:

bapclenet commented:
By the way, could we use 6LowPan without IPV6? I think I saw that you wanted to separate those layers.

authmillenon commented:
Yes, just remove the module from your `USEMODULE` list and use a device suitable for this packet format (ethernet (#2776) or slip (#2688) are currenly providing this).

bapclenet commented:
Great, thanks. looking forward to using all those new PRs :-)

OlegHahm commented:
@authmillenon, you were talking about IPv6 without 6LoWPAN, right?

authmillenon commented:
Yes.

haukepetersen commented:
now rebased, separated UDP shell command implementation and renamed the application to `ng_networking`. Now the application could be extended by anything that we add to the network stack (RPL, CoAP, etc).
Also the name must of course be changed, once (if) we give the network stack a name...

haukepetersen commented:
added 6lo fragmentation

bapclenet commented:
@haukepetersen, how could we change this PR in order to be able to select 6lowpan or UDP to send message? 
In shell_commands, add a 6low command and create the same file as udp.c for 6lowpan? (and we will do the same RPL. The commands could include init, send ...

haukepetersen commented:
I am not sure I understand what you mean... UDP sits on top of 6LoWPAN, so how will you OR between them?

authmillenon commented:
@bapclenet 6LoWPAN is an adaption layer between network and link layer, UDP is a transport layer protocol. What do you mean? If you want to de/activate 6LoWPAN use `ifconfig <if> [-]6lo`.

bapclenet commented:
Tell me if those solutions are possible:

1 | 2 | 3 | 4 | 5 
-----|-----|-----|-----|-----
UDP | UDP | UDP | UDP |  |
RPL | RPL | RPL | | | 
IPV6 | IPV6 | | |
6LowPan | | 6LowPan | 6LowPan | 6LowPan |
802.15.4 | 802.15.4 | 802.15.4 | 802.15.4 | 802.15.4 |

I mean that UDP, RPL, IPV6 and 6LowPan coulb be enable/disable easily.
Does it make sense? The idea is to be modular. @authmillenon, you told me that we can use 6LowPan without IPV6. UDP can run without RPL as well

gebart commented:
@bapclenet 
6lowpan is really not much more than a compression method for the ipv6 header and some fragmentation handling for the small MTU of IoT networks.
RPL is a routing protocol which uses the ICMPv6 protocol for passing information between routers.

I don't know about the internal structure of the ng stack but I guess 6lowpan depends on ipv6, @authmillenon can probably confirm/reject this statement.

In addition, the MTU of IEEE802.15.4 networks is 127 bytes, IPv6 requires a minimum of 1280 bytes MTU for proper conformance, so IPv6 without 6lowpan over a 802.15.4 link will probably not work in real life scenarios, but it might work in minimal examples.

UDP and RPL should be possible to disable independently. ICMPv6 is required for RPL though, just like UDP would be required for e.g. CoAP support.

authmillenon commented:
> I mean that UDP, RPL, IPV6 and 6LowPan could be enable/disable easily in this example.
Does it make sense? The idea is to be modular. @authmillenon, you told me that we can use 6LowPan without IPV6. UDP can run without RPL as well

1. RPL can be replaced by another routing protocol, yes. Otherwise you can only have one-hop 2. distances. 6LoWPAN can't be used independent without IPv6 and I never said it otherwise. It is the adaption layer for IPv6 over IEEE 802.15.4, so 6LoWPAN without IPv6 wouldn't make much sense.
2. IPv6 can however be used without 6LoWPAN, but only on Ethernet and SLIP (Serial-Line IP) devices.
3. On IEEE 802.15.4 devices you *need* 6LoWPAN because, as stated above, it is the adaption layer for IPv6 on this Link Layer.

You can use the modules independently and switch them out, you can even run them by themselves *for testing* (checking in- and output), but there is a limited number of combinations that make sense in a networking scenario. With whats currently available (including some outstanding PRs), these are (I left out the routing protocols since they are independent from the stack and should always be interchangable):

1 | 2 | 3 | 4 | 5 | 6 |
-----|-----|-----|-----|-----|-----|
UDP | UDP | UDP | | | |
IPv6 | IPv6 | IPv6 | IPv6 | IPv6 | IPv6 |
6LoWPAN | | | 6LoWPAN | | |
IEEE 802.15.4 | Ethernet | SLIP | IEEE 802.15.4 | Ethernet | SLIP |

You can also emulate IEEE 802.15.4 with ZEP (#2627) by tunneling IEEE 802.15.4 frames through UDP.

OlegHahm commented:
What about 6lo over SLIP (think of border router)?

authmillenon commented:
A border router would not send 6LoWPAN over SLIP, but IPv6. Otherwise it would be contrary to the point of being a border router between a LoWPAN and the Internet. It would use combination 4 for the LoWPAN side and combination 6 for the Internet side.

authmillenon commented:
If you want to send 6LoWPAN over anything else than IEEE 802.15.4 (or any of the currently drafted protocols), you need some kind of adaption layer (e. g. ZEP or the RIOT header in the old nativenet)

OlegHahm commented:
Right, I wasn't precise. Having an example with 6lo and SLIP is what I've meant.

bapclenet commented:
Thank you all for your clear explanations!

haukepetersen commented:
rebased.

authmillenon commented on examples/ng_networking/Makefile:
```diff
@@ -0,0 +1,29 @@
+
+# Specify the modules that are used
+USEMODULE += ng_netif
+USEMODULE += ng_sixlowpan_frag
```
This is not needed anymore

authmillenon commented on examples/ng_networking/Makefile:
```diff
@@ -0,0 +1,29 @@
+# Specify the modules that are used
+USEMODULE += ng_netif
+USEMODULE += ng_sixlowpan_frag
+USEMODULE += ng_ipv6
```
Use `ng_ipv6_default`

OlegHahm replied:
rather `ng_sixlowpan_default`, right?

authmillenon replied:
No… This example should also show-case non-IEEE802.15.4 devices. `ng_sixlowpan_default` is included automatically for IEEE802.15.4 devices since #2947, when `ng_ipv6_default` is present.

authmillenon commented on examples/ng_networking/Makefile:
```diff
@@ -0,0 +1,29 @@
+USEMODULE += ng_ipv6
+USEMODULE += ng_udp
+USEMODULE += ng_pktdump
+USEMODULE += uart0
```
Is this still needed?

haukepetersen replied:
on some boards yes. So I rather keep it in until we remove it completely from RIOT...

authmillenon commented:
Needs a README.md. The existence of the `USEMODUL += ps` line I see only as valid if the `ps` command is utilized in some way in that README ;-).

authmillenon commented:
I'm also wondering if the inclusion of `ng_icmpv6_echo` might be a good idea. It doesn't make the application code more complicated, but enriches its feature that automagically. @Lotterleben, what do you think?

OlegHahm commented:
:+1: for ICMP echo.

OlegHahm commented:
And what about `auto_init_ng_netif`?

OlegHahm commented:
And is there an equivalent to `defaulttransceiver` in gnrc?

Lotterleben commented:
@authmillenon what do you mean exactly by “inclusion of  ``ng_icmpv6_echo ``”? To  add some kind of “send_echo_req()” to bounce ICMP echo messages back and forth? Sorry to be a party pooper, but I'd be opposed to that too...
Assuming the examples are supposed to be both a tutorial (which imo should be focused on the one thing they're trying to explain) and a “clean slate” to start more advanced applications off of, I don't see why this application needs a richer feature set.
The above two use cases were what I've used the examples for, but if they are actually supposed to be something else, we should maybe make that more clear. ^^ (and my opinion on the matter may differ)

authmillenon commented:
> @authmillenon what do you mean exactly by “inclusion of ng_icmpv6_echo”? To add some kind of “send_echo_req()” to bounce ICMP echo messages back and forth? Sorry to be a party pooper, but I'd be opposed to that too...

No, just to add the module `ng_icmpv6_echo`. The rest (commands, reply behavior, etc) will be done by `shell_commands`, `auto_init` and `ng_icmpv6_echo` itself. The code for the application won't change. If we add a comment to the `USEMODULE += ng_icmpv6_echo` line, that it is optional, everybody that uses this application as a clean slate should be able to understand that they can omit the module.

Lotterleben commented:
Ooooh! Then, :+1: Maybe the optional USEMODULE += s could be grouped under a comment says that they add nifty shell functionality, but are optional? Just to be sure.

haukepetersen commented:
@authmillenon: I would like to leave `ps` in here, as IMHO the application is also used for testing and having just a quick look on stack usage is something quite useful...

haukepetersen commented:
adapted Makefile and rebased.

OlegHahm commented:
I think we still need a nice way to pull in a default transceiver for each (supported) platform.

authmillenon commented:
Doesn't https://github.com/RIOT-OS/RIOT/pull/3096 already proposes something in that direction?

OlegHahm commented:
Yes, it's just not implemented for all platforms.

haukepetersen commented:
How do we proceed with this PR? I have the feeling that the discussion about what to put in and what not could go on forever. So I would propose to merge this in the current state (so only do bug/style fixing for now) and adapt it in separate PRs to whatever we decide to have in. Does that work for everyone?

haukepetersen commented:
by the way, I removed the WIP status as I think we are past that stage...

PeterKietzmann commented:
Yes, makes sense to me. Until now I just tested the sender. As you also experienced, something with the port seems to behave strange when sniffing and looking at the data in wireshark.

haukepetersen commented:
This seems to be a bug in UDP, so probably unrelated to this PR?!

PeterKietzmann commented:
Well, I didn't prove if it's no bug, but transmitting data with my "personal" program I can see the correct port in wireshark

authmillenon commented on examples/ng_networking/Makefile:
```diff
@@ -0,0 +1,35 @@
+# Specify the mandatory networking modules for IPv6 and UDP
+USEMODULE += ng_ipv6_default
+USEMODULE += ng_udp
+# This application dumps received packets to STIO using the pktdump module
```
s/STIO/STDIO/

haukepetersen commented:
Hm, so this smells like a bug in this application... I will investigate...

authmillenon commented on examples/ng_networking/Makefile:
```diff
@@ -0,0 +1,35 @@
+USEMODULE += ng_udp
+# This application dumps received packets to STIO using the pktdump module
+USEMODULE += ng_pktdump
+# Additional networking modules that can be dropped if not needed
```
I know what this line means, but with my newbie glasses my only thought is "Huh? Which additional networking modules? Who can drop what?"

I think the simplest and clearest comment would just be

```Makefile
USEMODULE += ng_icmpv6_echo   # (optional, remove if ICMPv6 ping functionality is not wanted)
```

Lotterleben replied:
+1

OlegHahm commented:
I would also prefer that this application would out of the box which all supported boards, i.e. without touch the makefile.

haukepetersen commented:
fixed a byte-order bug in the dst port when sending and rebased.

haukepetersen commented:
@OlegHahm This application works out of the box with any board...

OlegHahm commented:
The last time I tried it didn't pull in the transceiver driver neither on native nor on IoT-LAB and I had to add `ng_at86rf231` or `ng_nativenet`. Don't have the time to check if this is fixed, but I will trust if you say it is.

haukepetersen commented:
I did not test on native, but on the iot-lab and the samr21 the radio is pulled in by default.

PeterKietzmann commented:
I confirmed the app works for iot-lab nodes

OlegHahm commented:
Confirmed: not pulling in the transceiver driver for native.

haukepetersen commented:
true, but you opened the PR yourself that will fix this, didn't you -> #3096

OlegHahm commented:
Yes, I did. So, I would either wait until #3096 is merged or prepare a temporary solution for this PR.

haukepetersen commented:
I'll wait.

haukepetersen commented:
rebased

haukepetersen commented:
rebased once more.

haukepetersen commented:
rebased on #3162. Now native also works out-of-the box.

haukepetersen commented:
adjusted to changes in #3162

PeterKietzmann commented:
I see no reason for not merging this PR after Travis is happy. Other opinions?

haukepetersen commented:
also rebased on #3188. Lets wait until #3162 and #3188 are merged, then we can merge this one

bapclenet commented:
@haukepetersen, have I tried it between two samr21? Long time ago, I couldn't make it work

authmillenon commented:
Needs rebase.

authmillenon commented:
@bapclenet just tested it with aeff85f merged into this. It worked for me with RIOT native <-> RIOT native, RIOT native <-> Linux, Linux <-> RIOT native, RIOT samr21-xpro <-> RIOT samr21-xpro, RIOT iot-lab_M3 <-> RIOT samr21-xpro, RIOT samr21-xpro <-> RIOT iot-lab_M3. So I say: let's merge this, as soon it was rebased, squashed and Travis is happy.

@haukepetersen that's an ACK. (there are some nitpicky things, but I will fix them when they still annoy me, when it was merged :-))

bapclenet commented:
@authmillenon, GREAT! Thanks for testing

authmillenon commented on examples/ng_networking/Makefile:
```diff
@@ -0,0 +1,35 @@
+# Include packages that pull up and auto-init the link layer.
+# NOTE: 6LoWPAN will be included if IEEE802.15.4 devices are present
+USEMODULE += ng_netif_default
+USEMODULE += auto_init_ng_netif
```
Is this still needed? Can't this be included automatically if `auto_init` and `ng_netif` are included?

haukepetersen replied:
As far as I understood: no.

PeterKietzmann replied:
Please test before taking it out. I'm not 100% sure but I think that my test application didn't work without this inclusion.

haukepetersen replied:
I leave it in to be safe for now, feel free to optimize later :-)

haukepetersen commented:
rebased and squashed. Lets wait for Travis, though.

haukepetersen commented:
Travis was not happy, as some boards did not have enough RAM for this application. Blacklisted them.

haukepetersen commented:
Travis is happy -> go, go, go!
